pkg/build/admission/admission.go

```diff
@@ -0,0 +1,78 @@
+package admission
+
+import (
+	"io"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/client"
+)
+
+type buildByStrategy struct {
+	*admission.Handler
+	client client.Interface
+}
+
+func init() {
+	admission.RegisterPlugin("BuildByStrategy", func(c kclient.Interface, config io.Reader) (admission.Interface, error) {
+		osclient := c.(client.Interface)
+		return NewBuildByStrategy(osclient), nil
+	})
+}
+
+// NewBuildByStrategy returns an admission control for builds that checks
+// on policy based on the build strategy type
+func NewBuildByStrategy(client client.Interface) admission.Interface {
+	return &buildByStrategy{
+		Handler: admission.NewHandler(admission.Create),
+		client:  client,
+	}
+}
+
+func (a *buildByStrategy) Admit(attr admission.Attributes) error {
+	if resource := attr.GetResource(); resource != "builds" {
+		return nil
+	}
+
+	build, ok := attr.GetObject().(*buildapi.Build)
+	if !ok {
+		return nil
+	}
+
+	return a.checkBuildAuthorization(attr.GetNamespace(), build)
+}
+
+func subjectAccessReview(build *buildapi.Build) *authorizationapi.SubjectAccessReview {
+	var resource string
+	switch build.Parameters.Strategy.Type {
+	case buildapi.DockerBuildStrategyType:
+		resource = "builds/docker"
+	case buildapi.CustomBuildStrategyType:
+		resource = "builds/custom"
+	case buildapi.SourceBuildStrategyType:
+		resource = "builds/source"
+	}
+
+	return &authorizationapi.SubjectAccessReview{
```

Comment:
This is checking whether the admission controller's client has permission to create a build pod with this strategy. Right now, admission plugins use the privileged client, which can do anything. Are you intending to check if the creating user has permission to do this?